api/odilo.py

```diff
@@ -0,0 +1,495 @@
+        ebook_streaming = False
+        if internal_format == 'EBOOK_STREAMING':
+            ebook_streaming = True
+            internal_format = 'EPUB'  # Try first ebook_streaming as EPUB
+
+        # Data just as 'x-www-form-urlencoded', no JSON
+        payload = dict(patronId=patron, format=internal_format)
+
+        response = self.patron_request(
+            patron, pin, self.CHECKOUT_ENDPOINT.replace("{recordId}", record_id),
+            extra_headers={'Content-Type': 'application/x-www-form-urlencoded'},
+            data=payload)
+
+        if response.content:
+            response_json = response.json()
+            if response.status_code == 404:
+                error_id, error_description = self.raise_exception_on_error(response_json)
+                raise CannotLoan('Error id: %s, description: %s' % (error_id, error_description))
+            else:
+                return self.loan_info_from_odilo_checkout(licensepool.collection, response_json)
+
+        elif response.status_code == 400 and ebook_streaming:
+            internal_format = 'PDF'  # Retry now with ebook_streaming as PDF
+            return self.checkout(patron, pin, licensepool, internal_format)
+
+        # TODO: we need to improve this at the API and use an error code
+        elif response.status_code == 400:
+            raise NoAcceptableFormat('record_id: %s, format: %s' % (record_id, internal_format))
+
+        raise CannotLoan('patron: %s, record_id: %s, format: %s' % (patron, record_id, internal_format))
+
+    def loan_info_from_odilo_checkout(self, collection, checkout):
+        start_date = self.extract_date(checkout, 'startTime')
+        end_date = self.extract_date(checkout, 'endTime')
+
+        return LoanInfo(
+            collection,
+            DataSource.ODILO,
+            Identifier.ODILO_ID,
+            checkout['id'],
+            start_date,
+            end_date,
+            checkout['downloadUrl']
+        )
+
+    def checkin(self, patron, pin, licensepool):
+        record_id = licensepool.identifier.identifier
+        loan = self.get_checkout(patron, pin, record_id)
+        url = self.CHECKIN_ENDPOINT.format(checkoutId=loan['id'], patronId=patron)
+
+        response = self.patron_request(patron, pin, url, method='POST')
+        if response.status_code == 200:
+            return response
+
+        error_code, error_description = self.raise_exception_on_error(response.json())
+        raise CannotReturn(error_code + ': ' + error_description)
+
+    @classmethod
+    def extract_date(cls, data, field_name):
+        if field_name not in data or not data[field_name]:
+            d = None
+        else:
+            # OdiloAPI dates are timestamps in milliseconds
+            d = datetime.datetime.utcfromtimestamp(float(data[field_name]) / 1000.0)
+        return d
+
+    @classmethod
+    def raise_exception_on_error(cls, data, custom_error_to_exception={}):
```

Comment:
A method called raise_exception_on_error should raise an exception if there's an error, and its return value should be irrelevant. You have code that calls this method and then raises an exception based on the return value. Either rename the method or (my preference) have it raise the exceptions itself.